src/native/managed/cdacreader/Microsoft.Diagnostics.DataContractReader.Contracts/ExecutionManagerHelpers/NibbleMap.cs

```diff
@@ -0,0 +1,284 @@
+// and an offset which we can then turn into the index of the nibble that covers that address.
+// If the nibble is non-zero, we have the start of a method and it is near the given address.
+// If the nibble is zero, we have to search backward first through the current map unit, and then through previous map
+// units until we find a non-zero nibble.
+//
+// For example (all code addresses are relative to some unspecified base):
+//  Suppose there is code starting at address 304 (0x130)
+//  Then the map index will be 304 / 32 = 9 and the byte offset will be 304 % 32 = 16
+//  Because addresses are 4-byte aligned, the nibble value will be 1 + 16 / 4 = 5  (we reserve 0 to mean no method).
+//  So the map unit containing index 9 will contain the value 0x5 << 24 (the map index 9 means we want the second nibble in the second map unit, and we number the nibbles starting from the most significant)
+//  Or 0x05000000
+//
+//  Now suppose we do a lookup for address 306 (0x132)
+//  The map index will be 306 / 32 = 9 and the byte offset will be 306 % 32 = 18
+//  The nibble value will be 1 + 18 / 4 = 5
+//  To do the lookup, we will load the map unit with index 9 (so the second 32-bit unit in the map) and get the value 0x05000000
+//  We will then shift to focus on the nibble with map index 9 (which again has nibble shift 24), so
+//  the map unit will be 0x00000005 and we will get the nibble value 5.
+//  Therefore we know that there is a method start at map index 9, nibble value 5.
+//  The map index corresponds to an offset of 288 bytes and the nibble value 5 corresponds to an offset of (5 - 1) * 4 = 16 bytes
+//  So the method starts at offset 288 + 16 = 304, which is the address we were looking for.
+//
+//  Now suppose we do a lookup for address 302 (0x12E)
+//  The map index will be 302 / 32 = 9 and the byte offset will be 302 % 32 = 14
+//  The nibble value will be 1 + 14 / 4 = 4
+//  To do the lookup, we will load the map unit containing map index 9 and get the value 0x05000000
+//  We will then shift to focus on the nibble with map index 9 (which again has nibble shift 24), so we will get
+//  the nibble value 5.
+//  Therefore we know that there is a method start at map index 9, nibble value 5.
+//  But the address we're looking for is map index 9, nibble value 4.
+//  We know that methods can't start within 32-bytes of each other, so we know that the method we're looking for is not in the current nibble.
+//  We will then try to shift to the previous nibble in the map unit (0x00000005 >> 4 = 0x00000000)
+//  Therefore we know there is no method start at any map index in the current map unit.
+//  We will then align the map index to the start of the current map unit (map index 8) and move back to the previous map unit (map index 7)
+//  At that point, we scan backwards for non-zero map units. Since there are none, we return null.
+
+internal class NibbleMap
```

Comment:
Should we think about putting version numbers on these classes. For instance, we know that there are more optimal forms of the NibbleMap that we might implement in the future, and I don't want to get rid of the existing code here, as it could be used to support an old version of the runtime. Do we want to think about calling these things NibbleMap_1 or something?

Reply:
yep.  we could do it now, or we could do it when we have a second algorithm.  `NibbleMap` doesn't leak out of the ExecutionManager contract, so we can cross that bridge when we get to it (also maybe we'll have better names than `_1` and `_2`)